### PR TITLE
Update routes while dragging

### PR DESCRIPTION
**Pull request by Miguel-Sanches**
This feature adds the capability of updating a route while dragging a waypoint or line.
The feature is disabled by default in the configuration and can be enabled by setting `updateWhileDrag` to true either through the config itself or through the public interface.

When dragging a line, a waypoint is automatically created (like `onDragUp` does) and the drag action is transferred from the line to this newly created waypoint.

The demo page has been edited as well, to include a checkbox to enable/disable this feature, in order to show it in action.

### CODE REVIEW

**Comment by smellyshovel** [comment]
@Miguel-Sanches Thanks for your contribution!
I've got a question to you. Maybe it would make sense to cancel the frequently-sent requests using the publicly-available abort controller? As far as I understand, you've implemented things in such a way so that each time there's a `mousemove` event a new request is sent. So maybe we should better check the next request and if the previous one isn't yet finished, we cancel it before sending the new one?

File: src/directions/main.ts [hunk 1]
```diff
@@ -587,6 +610,51 @@ export default class MapLibreGlDirections extends MapLibreGlDirectionsEvented {
     this.draw();
   }
 
+  protected async updateWhileDragging(e: MapMouseEvent | MapTouchEvent) {
+    /*
+     * letting other handlers know that updateWhileDragging is running
+     */
+    this._isCurrentlyUpdating = true;
```
Comment (smellyshovel) [inline src/directions/main.ts:617]: Why would we want to know in any other place whether it's updating or not?

**Comment by Miguel-Sanches** [comment]
Hi.

going to try and reply to all your concerns :)

There isn't an abort because the call to `fetchDirections` is on an await, making it so that no concurrent requests are made, as a new one can only be sent out after the previous one has finished.
But I see you already figured that out :)

As for the additional requests being fired even when the mouse is down but there is no movement, I think I have a solution, but I'd really like your input.

What if we were to use the same logic that you have for creating a new waypoint in `onDragUp` ?

If the new logic was encapsulated by the same if condition, no additional requests would be made, as far as I can see

`if (
      Math.abs(e.point.x - this.dragDownPosition?.x) >
        (this.configuration.dragThreshold >= 0 ? this.configuration.dragThreshold : 0) ||
      Math.abs(e.point.y - this.dragDownPosition?.y) >
        (this.configuration.dragThreshold >= 0 ? this.configuration.dragThreshold : 0)
    ) {`

Do you agree?

Thank you

**Comment by smellyshovel** [comment]
@Miguel-Sanches

> If the new logic was encapsulated by the same if condition, no additional requests would be made, as far as I can see
> `if ( Math.abs(e.point.x - this.dragDownPosition?.x) > (this.configuration.dragThreshold >= 0 ? this.configuration.dragThreshold : 0) || Math.abs(e.point.y - this.dragDownPosition?.y) > (this.configuration.dragThreshold >= 0 ? this.configuration.dragThreshold : 0) ) {`
> Do you agree?

I was about to suggest the same thing :)
Though there's a different concern arises.

Imaging the following scenario (with the `updateWhileDrag` option on).

The user starts dragging a waypoint (or a hoverpoint, doesn't matter here and forth). They started to hold the mouse button -> no requests are made (because of the `dragThreshold`). Then the waypoint is moved farther away than `dragThreshold`'s value. Now we start making requests one after another. Then the user stops their mouse, but doesn't release the button. No requests are made. And then they start moving the mouse again. **We should start making requests right away on this point, `dragThreshold` should not be taken into consideration this time**.

So I mean that dragThreshold should only be respected in comparison with the very initial waypoint's position, and not be nullified if the mouse stops/resumes its movement.

If that could be implemented, then I'm OK with your PR.

File: src/directions/types.ts [hunk 2]
```diff
@@ -265,6 +265,19 @@ export interface MapLibreGlDirectionsConfiguration {
    * ```
    */
   dragThreshold: number;
```
Comment (smellyshovel) [inline src/directions/types.ts:267]: Would be nice to update the `dragThreshold`'s docs as well, to indicate that this option also affects the behavior when your option is enabled.

Comment (Miguel-Sanches) [reply src/directions/types.ts:268]: done

File: demo/src/examples/1 User Interaction.svelte [hunk 3]
```diff
@@ -3,19 +3,32 @@
   import { examples } from "../router";
   import { location } from "svelte-spa-router";
   import AppSidebar from "../components/AppSidebar.svelte";
-  import maplibregl from "maplibre-gl";
+  import maplibregl, { Map } from "maplibre-gl";
```
Comment (smellyshovel) [inline demo/src/examples/1 User Interaction.svelte:6]: I specifically use exclusively namespaced maplibre's features. Even for types. Like `const map: maplibregl.Map`, but not `const map: Map`. Please, check the usages and change not to import namespaced features.

File: src/directions/main.ts [hunk 4]
```diff
@@ -49,13 +50,13 @@ export default class MapLibreGlDirections extends MapLibreGlDirectionsEvented {
   protected onDragMoveHandler: (e: MapMouseEvent) => void;
   protected onDragUpHandler: (e: MapMouseEvent) => void;
   protected onClickHandler: (e: MapMouseEvent) => void;
+  protected liveRefreshHandler: (e: MapMouseEvent | MapTouchEvent) => void;
```
Comment (smellyshovel) [inline src/directions/main.ts:53]: A good point on `MapTouchEvent`. Maybe makes sense to change accordingly the other occurances?

File: src/directions/main.ts [hunk 5]
```diff
@@ -98,14 +99,19 @@ export default class MapLibreGlDirections extends MapLibreGlDirectionsEvented {
   }
 
   protected async fetchDirections(originalEvent: MapLibreGlDirectionsWaypointEvent) {
+    this.abortController?.abort();
```
Comment (smellyshovel) [inline src/directions/main.ts:102]: Please, leave a comment why we need this. I know, but future contributors will get confused.

File: src/directions/main.ts [hunk 6]
```diff
@@ -362,16 +368,17 @@ export default class MapLibreGlDirections extends MapLibreGlDirectionsEvented {
     this.draw();
   }
 
-  protected dragDownPosition: {
-    x: number;
-    y: number;
-  } = {
+  protected dragDownPosition = {
     x: 0,
```
Comment (smellyshovel) [inline src/directions/main.ts:372]: Thanks!

File: src/directions/main.ts [hunk 7]
```diff
@@ -488,11 +521,28 @@ export default class MapLibreGlDirections extends MapLibreGlDirectionsEvented {
...
       this.hoverpoint.geometry.coordinates = [e.lngLat.lng, e.lngLat.lat];
     }
-
+    this.currentMousePosition = e.point;
     this.draw();
+
+    /*
+     * if the user selected a waypoint or routeLine and routes should update while dragging,
+     * we initiate the live updating process.
+     */
```
Comment (smellyshovel) [inline src/directions/main.ts:530]:
```
    /*
     * If the user selected a waypoint or a routeline and routes should update while dragging,
     * we initiate the live updating process.
     */
```

File: src/directions/main.ts [hunk 8]
```diff
@@ -488,11 +521,28 @@ export default class MapLibreGlDirections extends MapLibreGlDirectionsEvented {
...
+    if (this.configuration.refreshOnMove && !this.refreshOnMoveIsRefreshing) {
+      this.liveRefreshHandler(e);
+    }
   }
 
+  protected mouseMovementDetector: ReturnType<typeof setTimeout>;
```
Comment (smellyshovel) [inline src/directions/main.ts:536]: I'm not OK with the name. Sounds like some helper-method. Maybe something like `noMouseMovementTimer`? Or smtn similar...

File: src/directions/main.ts [hunk 9]
```diff
@@ -587,6 +639,48 @@ export default class MapLibreGlDirections extends MapLibreGlDirectionsEvented {
...
+        /*
+         * If a waypoint has been dragged, we fire a "movewaypoint" just like "onDragUp" does.
+         */
+        const waypointEvent = new MapLibreGlDirectionsWaypointEvent("movewaypoint", e, {
+          index: this._waypoints.indexOf(this.waypointBeingDragged),
+          initialCoordinates: this.waypointBeingDraggedInitialCoordinates,
+        });
+        this.fire(waypointEvent);
+        /*
+         * If the routing request has failed for some reason, restore the waypoint's original position.
```
Comment (smellyshovel) [inline src/directions/main.ts:671]: Are we restoring to position before dragDown happened? Or to the last while-moving-request's position?

File: src/directions/types.ts [hunk 10]
```diff
@@ -254,7 +254,8 @@ export interface MapLibreGlDirectionsConfiguration {
 
   /**
    * A minimal amount of pixels a waypoint or the hoverpoint must be dragged in order for the drag-event to be
-   * respected. Should be a number >= `0`. Any negative value is treated as `0`.
+   * respected, and for network requests to be made when using refreshOnMove = true. Should be a number >= `0`.
```
Comment (smellyshovel) [inline src/directions/types.ts:257]: Change please

> when using refreshOnMove = true.

to something like (check out similar places)

when {@link refreshOnMove|`refreshOnMove`} equals `true`.

File: src/directions/types.ts [hunk 11]
```diff
@@ -265,6 +266,19 @@ export interface MapLibreGlDirectionsConfiguration {
...
+
+  /**
+   * Whether to update a route while dragging a waypoint/hoverpoint instead of only when dropping it
+   *
+   * @default `false`
+   *
+   * @example
+   * ```
+   * // make the route update while dragging
+   * updateWhileDrag: true
```
Comment (smellyshovel) [inline src/directions/types.ts:278]: Rename please :)

**Review by smellyshovel — CHANGES_REQUESTED** [review]

File: src/directions/types.ts (hunk 10 above)
Comment (smellyshovel) [reply src/directions/types.ts:257]: This is how TypeDoc works. Some magic and not a very clear comment, but turns out into a nicely-formatted documentation.

File: demo/src/examples/1 User Interaction.svelte (hunk 3 above)
Comment (Miguel-Sanches) [reply demo/src/examples/1 User Interaction.svelte]: changed in this file and updated the already existing third example page to reflect the same methodology

File: demo/src/examples/1 User Interaction.svelte [hunk 12]
```diff
@@ -3,19 +3,32 @@
...
   import "maplibre-gl/dist/maplibre-gl.css";
   import style from "../assets/map/style/style.json?url";
   import MapLibreGlDirections from "@maplibre/maplibre-gl-directions";
 
   const meta = examples.find((example) => example.path === $location);
 
   let mapRef: HTMLElement | undefined = undefined;
+  let map: Map | undefined = undefined;
```
[the remark this answers is not in the document]
Comment (Miguel-Sanches) [reply demo/src/examples/1 User Interaction.svelte]: changed in this file and updated the already existing third example page to reflect the same methodology

File: src/directions/main.ts (hunk 4 above)
Comment (Miguel-Sanches) [reply src/directions/main.ts:53]: Added `MapTouchEvent` to pre-existing handlers

File: src/directions/main.ts [hunk 13]
```diff
@@ -49,13 +50,13 @@ export default class MapLibreGlDirections extends MapLibreGlDirectionsEvented {
...
   protected onDragUpHandler: (e: MapMouseEvent) => void;
   protected onClickHandler: (e: MapMouseEvent) => void;
+  protected liveRefreshHandler: (e: MapMouseEvent | MapTouchEvent) => void;
 
   protected _waypoints: Feature<Point>[] = [];
   protected snappoints: Feature<Point>[] = [];
   protected routelines: Feature<LineString>[][] = [];
   protected selectedRouteIndex = 0;
   protected hoverpoint: Feature<Point> | undefined = undefined;
-
```
[the remark this answers is not in the document]
Comment (Miguel-Sanches) [reply src/directions/main.ts]: IDE auto formatting. sorry. I've added it back

File: src/directions/main.ts (hunk 8 above)
Comment (Miguel-Sanches) [reply src/directions/main.ts]: Changed

File: src/directions/main.ts (hunk 9 above)
Comment (Miguel-Sanches) [reply src/directions/main.ts:676]: To the position before dragDown happened

**Review by smellyshovel — APPROVED** [review]

**Pull request merged**